src/main/java/gregtech/api/recipes/map/MapItemStackIngredient.java

```diff
@@ -1,30 +1,46 @@
 package gregtech.api.recipes.map;
 
+import gregtech.api.recipes.ingredients.GTRecipeInput;
+import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import net.minecraft.item.ItemStack;
 import net.minecraft.nbt.NBTTagCompound;
 
+import java.util.Collection;
+
 public class MapItemStackIngredient extends AbstractMapIngredient {
 
+
     protected ItemStack stack;
     protected int meta;
     protected NBTTagCompound tag;
+    protected GTRecipeInput gtRecipeInput = null;
 
-    public MapItemStackIngredient(ItemStack stack) {
-        this.stack = stack;
-        this.meta = stack.getMetadata();
-        this.tag = stack.getTagCompound();
-    }
     public MapItemStackIngredient(ItemStack stack, int meta, NBTTagCompound tag) {
         this.stack = stack;
         this.meta = meta;
         this.tag = tag;
     }
 
+    public MapItemStackIngredient(ItemStack stack, GTRecipeInput gtRecipeInput) {
+        this.stack = stack;
+        this.meta = stack.getMetadata();
+        this.tag = stack.getTagCompound();
+        this.gtRecipeInput = gtRecipeInput;
+    }
+
+    public static Collection<AbstractMapIngredient> from(GTRecipeInput r) {
+        ObjectArrayList<AbstractMapIngredient> list = new ObjectArrayList<>();
+        for (ItemStack s : r.getInputStacks()) {
+            list.add(new MapItemStackIngredient(s,r));
+        }
+        return list;
+    }
+
     @Override
     public boolean equals(Object o) {
         if (super.equals(o)) {
             MapItemStackIngredient other = (MapItemStackIngredient) o;
-            return stack.getItem() == other.stack.getItem() && meta == other.meta && ItemStack.areItemStackTagsEqual(stack, other.stack);
+            return other.gtRecipeInput.acceptsStack(this.stack);
```

Comment:
Wouldn't this always fail if the first constructor is used, as `gtRecipeInput` would not be initialized?

Reply:
we will only be matching MapItemStackIngredients made from the itemstacks in the inventory against those made from GTrecipeInputs